continuing to isolate v2 hashing code

aryairani opened the pull request:
- [x] let's fix test failures in `topic/rehash-codebase` before merging this; apart from that, it's okay to review now
- [ ] `gitsync22.sc.gist` is failing due to unknown explicitly referenced hash `frj8ob9ugr`
- [ ] `gitsync22.sc.history` is failing due to unknown explicitly referenced hash `0u7no051k7`

This PR:
1. moves Causal data structures from `Unison.Causal` to `Unison.Causal.Type` to avoid cyclic dependencies with hashing module, leaving just convenience functions and re-exports in the original file.
2. deletes Causal cachedRead / deserialization stuff; (this could potentially be a separate PR, but doesn't seem controversial)
3. delegates all Causal hashing to `Hashing.hashCausal`. I created new read-only patterns `One` / `Cons` / `Merge`, and rename the constructors to `UnsafeOne`, `UnsafeCons`, `UnsafeMerge`; these are used now in very limited circumstances.
4. moves most of `Hashable.hs` into `BuildHashable.hs`; Hashable is less composable now by design.
5. leaves `Unison.Hashable` in place for some bit sprinkling that Parser.hs is still doing, but removed the Hashable1 typeclass from that module. Some arbitrary external hashing library could be used here, but I didn't bother to find and integrate one.

Commit:
merge trunk and 2625

Commit:
Merge branch 'topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

# Conflicts:
#	unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
#	unison-cli/src/Unison/CommandLine/OutputMessages.hs

Commit:
pull Hashable into Hashing.V2 internals and out of Causal

Commit:
Merge branch 'topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/Convert.hs:
```diff
@@ -347,44 +347,47 @@ m2hPatch (Memory.Patch.Patch termEdits typeEdits) =
+hashBranch0 :: Memory.Branch.Branch0 m -> Hash
+hashBranch0 = Hashing.Branch.hashBranch . m2hBranch0
 
 -- hashing of branches isn't currently delegated here, because it's enmeshed
```
Are we planning to change this at all as part of this migration?

aryairani replied:
Oops, yes, this PR would change this, but I forgot to change the comment.

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/Hashable.hs:
```diff
@@ -0,0 +1,139 @@
+instance (Hashable d1, Hashable d2, Hashable d3, Hashable d4) => Hashable (Relation4 d1 d2 d3 d4) where
+  tokens s = [accumulateToken $ Relation4.toNestedList s]
+
+class Functor f => Hashable1 f where
```
I know you're just moving this, but does this actually need the Functor constraint? Seems like it doesn't depend on it, and the implementer of the instance should probably be able to decide their own constraints right?

I don't mind if it stays, mostly just asking out of curiousity.

aryairani replied:
Seems like an oversight and that it should go.

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/Branch.hs:
```diff
@@ -21,6 +20,11 @@ newtype MdValues = MdValues (Set MetadataValue)
 
+newtype NameSegment = NameSegment Text deriving (Eq, Ord, Show)
+
+hashBranch :: Raw -> Hash
```
Maybe `hashRawBranch` to avoid confusion with the other `hashBranch` and `hashBranch0`?

aryairani replied:
Hm I think it's okay, this is part of a pretty different interface, and I wouldn't anticipate ever importing both packages.

But also, the fact that the type defined in this file is called `Raw` is something of a historical artifact, and could just as easily be named `Branch` as well. I think the function here is named more after the module and the concept than the data type.

ChrisPenner reviewed (commented):
Looks fine to me; I'll admit I don't love having overlapping names, even in different modules, generally I find it just makes things a bit tougher to grok at a glance, but that'd be my only criticism 😄

aryairani commented on stack.yaml:
```diff
@@ -44,7 +44,7 @@ extra-deps:
 
 ghc-options:
  # All packages
- "$locals": -Wall -Werror -Wno-name-shadowing -Wno-type-defaults -Wno-missing-pattern-synonym-signatures -fprint-expanded-synonyms -fwrite-ide-info #-freverse-errors
```
Whoops, didn't mean to check in this change.

aryairani commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -495,7 +495,6 @@ loop = do
           updateAtM = Unison.Codebase.Editor.HandleInput.updateAtM inputDescription
           unlessGitError = unlessError' Output.GitError
           importRemoteBranch ns mode = ExceptT . eval $ ImportRemoteBranch ns mode
-          viewRemoteBranch ns = ExceptT . eval $ ViewRemoteBranch ns
```
Not sure why this shows up here, I thought someone else deleted it.

Commit:
misc PR feedback

Commit:
Merge remote-tracking branch 'origin/topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

Commit:
update checklist

Commit:
Merge remote-tracking branch 'origin/topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

Commit:
un-un-generalize Causal.hs functions, and make Hashable less extensible

It may be worth it to un-split it from Causal.Type.hs also.

Moved Hashable.hs guts into BuildHashable.hs

un-un-specialize Unison.Hashable.hs

un-delete Unison.Test.Codebase.Causal.hs (checked out from #2690)

Commit:
build Causal hashes in Unison.Hashing.V2.Convert as Hashing

Commit:
just have one path through causal construction

Commit:
Merge remote-tracking branch 'origin/topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

# Conflicts:
#	parser-typechecker/src/Unison/Codebase/Branch.hs
#	parser-typechecker/src/Unison/Hashing/V2/Convert.hs

Commit:
restore causal test suite

Commit:
remove `Unison.Hashable.Hashable1`

aryairani commented on unison-core/src/Unison/Hashable.hs:
```diff
@@ -37,9 +37,9 @@ class Accumulate h where
-accumulate' :: (Accumulate h, Hashable t) => t -> h
-accumulate' = accumulate . (hashVersion :). tokens
-  where hashVersion = Tag 2
+hash, accumulate' :: (Accumulate h, Hashable t) => t -> h
+accumulate' = accumulate . tokens
+hash = accumulate'
```
This is no longer used for our hashing, so we don't need to salt anything.  Perhaps we'll refactor `Parser.hs` at some point to no longer require this module.

aryairani commented on parser-typechecker/src/Unison/Hashing/V2/Hashable.hs:
```diff
@@ -0,0 +1,21 @@
+
+instance BuildHashable.Hashable a => Hashable [a] where
+  hash = BuildHashable.hash
+instance BuildHashable.Hashable a => Hashable (Set a) where
```
```suggestion

instance BuildHashable.Hashable a => Hashable (Set a) where
```

aryairani committed:
Update parser-typechecker/src/Unison/Hashing/V2/Hashable.hs

Commit:
comments

Commit:
Merge branch 'topic/rehash-codebase-reexport-hash' of ghs:unisonweb/unison into topic/rehash-codebase-reexport-hash

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal.hs:
```diff
@@ -297,14 +136,25 @@ stepDistinctM
 
+-- | Causal construction should go through here for uniformity;
+-- with an exception for `one`, which avoids an Applicative constraint.
+fromList :: (Applicative m, Hashable e) => e -> [Causal m h e] -> Causal m h e
```
The name `fromList` isn't really suggesting how this should be used to me.
It's also a bit weird to have a `fromList` that takes an extra argument than just the list.

```suggestion
consOntoTails :: (Applicative m, Hashable e) => e -> [Causal m h e] -> Causal m h e
```

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal.hs:
```diff
@@ -297,14 +136,25 @@ stepDistinctM
+fromList :: (Applicative m, Hashable e) => e -> [Causal m h e] -> Causal m h e
+fromList e (List.nubOrdOn currentHash -> tails) = case tails of
+  [] -> UnsafeOne h e
+  t : [] -> UnsafeCons h e (tailPair t)
```
What should we do here if `h` is equal to `currentHash t`?
We should probably de-dupe this right?

aryairani replied:
No, we do that in `consDistinct`, but not in `cons`.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal/Type.hs:
```diff
@@ -0,0 +1,139 @@
+import Prelude hiding (head, read, tail)
+
+{-
+`Causal a` has 5 operations, specified algebraically here:
```
Love the descriptions, but maybe these should just be haddocks (at the very least a module doc, so it shows up in haddocks?)
I'm worried about this description, or the contained types, going out of date.
Case-in-point, the `Causal` type in the descriptions switches between `Causal a` and `Causal m a`, when in fact the type is `Causal m h a` 🤔 

---

Edit: after reading through this module I see that none of these operations exist as combinators here as the docs seem to imply, are you referring to the Patterns? Where is `sequence` defined?

Now I'm very confused 😓 ; not sure what this block of docs is trying to say.

---

Edit 2:
I didn't realize this was simply moved (in the future, notes within the PR about what changed, what moved, etc. really help with this 😄 )
I think the comments still stand, but up to you whether we make those changes now or not.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal/Type.hs:
```diff
@@ -0,0 +1,139 @@
+* `one : a -> Causal m a`, satisfying `head (one hd) == hd`
+* `cons : a -> Causal a -> Causal a`, satisfying `head (cons hd tl) == hd` and
+          also `before tl (cons hd tl)`.
+* `merge : CommutativeSemigroup a => Causal a -> Causal a -> Causal a`, which is
```
Not sure what this is implying; The fact that this is a type signature implies the existence of a `CommutativeSemigroup` class somewhere which I can't find in this PR. I suspect you mean it more conceptually, e.g. "You need to tell me how to merge two or more branches in order to merge", but I think I'd rather have that written in English rather than referring to classes we don't actually use.

I don't see which definition this is referring to in this module though, nothing seems to accept anything resembling a CommutativeSemigroup.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal/Type.hs:
```diff
@@ -0,0 +1,139 @@
+{-# COMPLETE One, Cons, Merge #-}
+
+
+Lens.makeLensesFor [("head", "head_")] ''Causal
```
This lens is unsafe; editing the head invalidates the hash, but doesn't update it.

We can either downgrade this to a 'getter', or replace with the appropriate lens that updates the hash after updating.

aryairani replied:
We need to merge in the change from Causal.hs and/or do this split off of trunk first and see what happens.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal/Type.hs:
```diff
@@ -0,0 +1,139 @@
+
+Lens.makeLensesFor [("head", "head_")] ''Causal
+
+children :: Causal m h e -> Seq (m (Causal m h e))
```
```suggestion
predecessors :: Causal m h e -> Seq (m (Causal m h e))
```

Not sure whether now is the right time to change it or not (I'll leave that up to you), but I believe we agreed `children` isn't the best name for this concept.

aryairani replied:
We think this is also already changed on trunk.

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Causal.hs:
```diff
@@ -35,181 +31,33 @@ where
-
--- h is the type of the pure data structure that will be hashed and used as
--- an index; e.g. h = Branch00, e = Branch0 m
-data Causal m h e
```
Is it necessary to split this module up right now? Seems like it's likely to cause us merge-conflict issues, would it be possible to do this refactor on trunk and merge it in instead?
Refactoring in a long-lived feature branch can cause a lot of headaches sometimes 😅

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:
```diff
@@ -0,0 +1,140 @@
+  tokens = map accumulateToken
+
+instance (Hashable a, Hashable b) => Hashable (a, b) where
+  tokens (a, b) = [accumulateToken a, accumulateToken b]
```
I notice we have a lot of overlap between our Hashables; e.g. AFAICT all of these distinct types would have the same hash:

```
[1, 2]
(1, 2)
Set.fromList [1, 2]
```
etc.;

Should we be tagging these somehow? Or maybe prepending a `Text "Set"` to the front or something to disambiguate?

Maybe it's fine with the current implementation of things, but seems like it might be a really subtle bug just waiting to happen 😅

aryairani replied:
We talked on the phone and decided the tags belong on the instances for sum types, and not necessarily to distinguish between multiple instances.

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:
```diff
@@ -0,0 +1,140 @@
+instance (Hashable d1, Hashable d2, Hashable d3, Hashable d4) => Hashable (Relation4 d1 d2 d3 d4) where
+  tokens s = [accumulateToken $ Relation4.toNestedList s]
+
+class Hashable1 f where
```
Seems kinda weird to have this class in the middle of a bunch of Hashable instances; also do we not actually define any Hashable1 instances?
I'd expect to see instances for basic container types here if we actually used it.

aryairani replied:
We define them on the other side — ABT / Term / Type / DataDeclaration, etc.

ChrisPenner commented on parser-typechecker/src/Unison/Hashing/V2/BuildHashable.hs:
```diff
@@ -0,0 +1,140 @@
+module Unison.Hashing.V2.BuildHashable where
```
IIRC we wanted to only export certain things for safety, not sure quite where this implementation ended up on that,
was there any plan to restrict exports here?

ChrisPenner reviewed (commented):
I left a bunch of comments, but I'll admit that I'm confused about what I'm reviewing here, and I think I ended up reviewing a bunch of code that was either just moved or was copy-pasted, a tour of this PR would be helpful for me 😄

aryairani commented on parser-typechecker/src/Unison/Codebase/Causal.hs:
```diff
@@ -297,14 +136,25 @@ stepDistinctM
   => (e -> n e) -> Causal m h e -> n (Causal m h e)
 stepDistinctM f c = (`consDistinct` c) <$> f (head c)
 
+-- | Causal construction should go through here for uniformity;
+-- with an exception for `one`, which avoids an Applicative constraint.
```
Update docs to mention that the list is de-duped.

aryairani commented on parser-typechecker/src/Unison/Hashing/V2/Patch.hs:
```diff
@@ -2,17 +2,20 @@
 import Unison.Hashing.V2.TermEdit (TermEdit)
 import Unison.Hashing.V2.TypeEdit (TypeEdit)
 
+hashPatch :: H.Accumulate h => Patch -> h
+hashPatch = H.accumulate'
```
Monomorphize references to Accumulate

ChrisPenner commented:
We decided to try to get tests passing on topic/rehash before merging this. Mitchell and I will try to accomplish that while Arya is on holidays.

Commit:
monomorphised Patch.hashPatch and Causal.hashCausal

mitchellwrosen committed:
⅄ topic/rehash-codebase → topic/rehash-codebase-reexport-hash

mitchellwrosen commented:
~@ChrisPenner and I are getting ready to merge this, but one mystery remains. Actually, it remains a mystery whether it's a mystery.~

~On this branch, I discovered after a lot of printf debugging that we are actually changing how we hash branches (`Branch0`), relative to what's in `topic/rehash-codebase` right now.~

~Is that expected?~

mitchellwrosen commented:
I did a little bit more debugging, and it turns out the actual problem is we are hashing a `One` causal differently.

Previously, we just used the branch hash directly. Now, we pass it and the empty list of predecessors along to the underlying hashable instance:

```haskell
instance H.Hashable Causal where
  tokens c = H.tokens $ branchHash c : Set.toList (parents c)
```

It seems fine to tumble in the hash of the empty list to a `One` causal, I just wanted to call that out. It's also easy to get the old behavior:

```haskell
instance H.Hashable Causal where
  tokens c =
    case Set.toList (parents c) of
      [] -> H.tokens [branchHash c]
      cs -> H.tokens (branchHash c : cs)
```

ChrisPenner committed:
Merge branch 'topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

Commit:
Merge branch 'topic/rehash-codebase' into topic/rehash-codebase-reexport-hash

# Conflicts:
#	parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

ChrisPenner committed:
Merge branch 'topic/rehash-codebase' into topic/rehash-codebase-reexport-hash
